api/src/main/java/org/openmrs/module/pihcore/setup/ConfigurationSetup.java

```diff
@@ -139,6 +140,7 @@ public void setupBase() throws Exception {
         removeOldPrivileges(); // TODO: This can likely be removed altogether at this point, or moved to liquibase
         ReportSetup.cleanupOldReports(); // TODO: could move this to liquibase or to a fixed scheduled task
         initializerMessageSource.addFallbackLanguage("ht", "fr");
+        AuthenticationConfig.registerClassLoader(getClass().getClassLoader());
```

Comment:
Because the authentication module is fully configurable via runtime properties in such a way that all of the classes are instantiated and configured by the authentication module itself, and not injected from other modules (eg. via Spring, or other patterns), I added the capability to register additional class loaders with the authentication module that it can use to find and create classes.  @mogoodrich I'd be interested in your thoughts and whether there are pitfalls to this approach or if you think it is generally reflective of a poor design.

Reply:
Correct @mogoodrich .  The Authentication module uses an OpenMRS ModuleClassLoader instance, which has access to OpenMRS classes, and any module classes that it indicates are required modules or aware-of modules.  Since authentication is not aware of pihcore, it is not able to load a custom PIH authentication scheme by className.  So the solution I introduced was to allow downstream modules (eg. PIH core) to register classloaders with the authentication module, so that when the authentication module loads classes by name, it can use these classloaders to accomplish this.

An alternative would be to introduce a factory pattern (which is similar to the above, but abstracted away from the classloader level) and allow downstream modules (eg. PIH Core) to inject a custom factory to use to load specific classes.  For example, we could have an `AuthenticationSchemeFactory` interface in the authentication module, and then we could have an implementation of this in PIH Core, and then we could register this custom AuthenticationSchemeFactory via Spring autowiring or explicitly in the PIH Core Activator.

```
interface AuthenticationSchemeFactory {
    AuthenticationScheme newAuthenticationScheme(String schemeId);
}
```

This is possibly the cleaner approach?  Just requires an additional interface, and an additional custom class in each downstream module, and is less flexible - i.e. we'd need other factories for producing other types of custom classes - which is why I didn't go this route.

Happy to discuss and decide which to run with.

Reply:
Happy to discuss further if you like, but also happy to run with whichever of the approaches you think it best... I just was trying to make sure I understood how it worked.

Reply:
Actually, maybe just add a comment here explaining what this is for?

Reply:
done